### PR TITLE
SNOW-1707419: Improve error message for unknown timezone in tz_convert and tz_localize

### PR DESCRIPTION
**Pull request by sfc-gh-helmeleegy**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1707419

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.

3. Please describe how your code solves the related issue.

   Improve error message for unknown timezone in tz_convert and tz_localize.

### CODE REVIEW

File: tests/integ/modin/frame/test_tz_convert.py [hunk 1]
```diff
@@ -101,7 +102,7 @@ def test_tz_convert_negative(axis, level, copy, exception):
 
     with pytest.raises(exception):
         snow_df.tz_convert(
-            tz="UTC",
+            tz=tz,
```
Comment (sfc-gh-lmukhopadhyay) [inline tests/integ/modin/frame/test_tz_convert.py:105]: Maybe we can match the error message in all these negative tests with something like
```
pytest.raises(
            exception, match=".*method doesn't support \'tz={tz}\'"
):
```

**Review by sfc-gh-lmukhopadhyay — CHANGES_REQUESTED** [review]
LGTM! I just had one suggestion for checking the error message in the negative tests

**Review by sfc-gh-azhan — COMMENTED** [review]
why not just support it?

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py [hunk 2]
```diff
@@ -17294,6 +17295,10 @@ def dt_tz_localize(
             ErrorMessage.parameter_not_implemented_error("ambiguous", method_name)
         if not isinstance(nonexistent, str) or nonexistent != "raise":
             ErrorMessage.parameter_not_implemented_error("nonexistent", method_name)
+        if isinstance(tz, str) and tz not in pytz.all_timezones:
```
Comment (sfc-gh-jjiao) [inline src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:17298]: are we saying we do not support other format for timezone string that are not in pytz.all_timezones? 

pytz.all_timezones does not support anything for UTC...

Comment (sfc-gh-azhan) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:17298]: Both pandas and Snowflake are supporting "UTC+09:00", why we cannot support it here?

**Review by sfc-gh-jjiao — CHANGES_REQUESTED** [review]
I am not sure why we are checking the input string against pytz.all_timezones. 

By comparison, native pandas does support timezone format like UTC:+09:00 etc

Screenshot 2024-10-11 at 11 52 56 AM

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py (hunk 2 above)
Comment (sfc-gh-helmeleegy) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:17298]: @sfc-gh-azhan where did see that Snowflake accepts taking "UTC+09:00" as an input timezone?

When I run a query like:
```
select convert_timezone('Europe/Warsaw', 'UTC+09:00', '2024-01-01 00:00:00'::TIMESTAMP_NTZ)
```
I get back:
```
Unknown timezone: 'UTC+09:00'
```

Comment (sfc-gh-azhan) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:17298]: Oh I guess my memory was wrong. Then I think we are good to return NotImplementedError for now.

File: tests/integ/modin/frame/test_tz_convert.py (hunk 1 above)
Comment (sfc-gh-helmeleegy) [reply tests/integ/modin/frame/test_tz_convert.py:105]: The issue is that the `NotImplementedError` here can be caused by different scenarios of unsupported input parameters, not just `tz`. That's why we only check for the error type and not the error message.

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py (hunk 2 above)
Comment (sfc-gh-helmeleegy) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:17298]: Sounds good. We can update the docs to reflect this in a followup PR. Does this work?

**Comment by sfc-gh-helmeleegy** [comment]
> I am not sure why we are checking the input string against pytz.all_timezones.
> 
> By comparison, native pandas does support timezone format like UTC:+09:00 etc
> 
> Screenshot 2024-10-11 at 11 52 56 AM

It's mainly to stay consistent with Snowflake's restrictions on input timezones. Please refer to the previous thread.

**Comment by sfc-gh-jjiao** [comment]
> > I am not sure why we are checking the input string against pytz.all_timezones.
> > By comparison, native pandas does support timezone format like UTC:+09:00 etc
> > Screenshot 2024-10-11 at 11 52 56 AM
> 
> It's mainly to stay consistent with Snowflake's restrictions on input timezones. Please refer to the previous thread.

Sounds good!

This PR is only dealing with error message?   

```
            ErrorMessage.not_implemented(
                f"Snowpark pandas 'tz_convert' method doesn't support 'tz={tz}'"
            )
```
Could we update this PR directly to raise NotImplementedError  or you are going to merge this first and then create a new PR to do that?

**Comment by sfc-gh-helmeleegy** [comment]
> 
> Sounds good!
> 
> This PR is only dealing with error message?
> 
> ```
>             ErrorMessage.not_implemented(
>                 f"Snowpark pandas 'tz_convert' method doesn't support 'tz={tz}'"
>             )
> ```
> 
> Could we update this PR directly to raise NotImplementedError or you are going to merge this first and then create a new PR to do that?

This PR already raises NotImplementedError.

**Review by sfc-gh-lmukhopadhyay — APPROVED** [review]

**Comment by sfc-gh-helmeleegy** [comment]
> why not just support it?

Addressed in other threads.

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py [hunk 3]
```diff
@@ -23,6 +23,7 @@
 import pandas.core.resample
 import pandas.io.parsers
 import pandas.io.parsers.readers
+import pytz  # type: ignore
```
[the remark this answers is not in the document]
Comment (sfc-gh-helmeleegy) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:26]: `mypy` would complain with:
```
error: Library stubs not installed for "pytz"
```

**Review by sfc-gh-jjiao — APPROVED** [review]

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sfc-gh-azhan) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:26]: can you add `types-pytz` into https://github.com/snowflakedb/snowpark-python/blob/e076031d35c169b779df37bdd18dfd3cae78c20d/.pre-commit-config.yaml#L101

**Pull request merged**